ReactiveCocoa/Swift/Signal.swift

```diff
@@ -168,25 +156,32 @@ public func take<T, E>(count: Int)(signal: Signal<T, E>) -> Signal<T, E> {
 	precondition(count >= 0)
 
 	return Signal { observer in
+		if count == 0 {
+			sendCompleted(observer)
+			return nil
+		}
+
 		var taken = 0
 
-		return signal.observe(next: { value in
-			if taken < count {
-				taken++
-				sendNext(observer, value)
-			} else {
-				sendCompleted(observer)
+		return signal.observe(Signal.Observer { event in
+			switch event {
+			case let .Next(value):
+				if taken < count {
+					taken++
+					fallthrough
```

Comment:
In this context, I think it makes sense (although maybe I should be restructuring the code to make it clearer).

The idea is, “if we've not exhausted our count, use the default behavior of passing everything through.” The inverse applies to `skip` too.